File: CPP/ByRound/1891/1891_C_Smilo_and_Monsters.cpp
// Problem: CF 1891 C - Smilo and Monsters
// https://codeforces.com/contest/1891/problem/C

/*
Algorithm: Greedy with Deque
Time Complexity: O(n log n) due to sorting
Space Complexity: O(n) for the deque storage

This solution uses a greedy approach to minimize the number of attacks.
We sort the horde sizes and use a deque to efficiently access both ends.
The combo counter 'com' tracks how many monsters we can eliminate with ultimate attack.
We process the smallest horde first, using regular attacks to build up combo,
and use ultimate attack when beneficial to reduce total attack count.
*/
#include <algorithm>
#include <deque>
#include <iostream>

using namespace std;
int t, n;
void solve() {
  cin >> n;
  deque<int> deq(n);
  for (int &e : deq)
    cin >> e;
  sort(deq.begin(), deq.end()); // Sort the horde sizes
  int com = 0; // Combo counter
  long long sum = 0; // Total number of attacks
  while (deq.size()) { // Continue until all hordes are cleared
    if (deq.size() == 1) { // If only one horde left
      if (deq.front() == 1) {
        sum++; // One regular attack
        break;
      }
      if (deq.front()) // If there are monsters in the last horde
        sum++; // One regular attack to reduce it to 1
      if (com < deq.front()) // If combo can help reduce attack count
        sum += (deq.front() - com + 1) / 2; // Calculate minimum attacks needed
      deq.pop_back(); // Remove the last horde
    } else { // Multiple hordes remain
      if (com + deq.front() < deq.back()) { // If combo+smallest < largest
        sum += deq.front(); // Attack smallest horde completely
        com += deq.front(); // Increase combo by number of monsters attacked
        deq.pop_front(); // Remove smallest horde
      } else { // If combo is sufficient to trigger ultimate attack
        deq.front() -= (deq.back() - com); // Reduce monsters in smallest horde
        sum += deq.back() - com + 1; // Count of attacks to use ultimate
        com = 0; // Reset combo counter
        deq.pop_back(); // Remove largest horde
      }
    }
  }
  cout << sum << '\n';
}
int main() {
  cin.tie(0)->sync_with_stdio(0);
  cin >> t;
  while (t--)
    solve();
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/C/1891_C_Smilo_and_Monsters.py
# Problem: CF 1891 C - Smilo and Monsters
# https://codeforces.com/contest/1891/problem/C

"""
Algorithm: Greedy approach to minimize number of attacks.
Techniques: Sorting, two-pointer like logic.

Time Complexity: O(n log n) due to sorting.
Space Complexity: O(1) if we ignore the space used for input.

The idea is to use a greedy strategy where we prioritize using 
the second type of attack (ultimate attack) on the largest hordes 
to maximize the number of monsters destroyed per attack. This reduces 
the total number of attacks needed.
"""

for _ in range(int(input())):  # Read number of test cases
    n = int(input())  # Read number of hordes
    a = list(map(int, input().split()))  # Read monsters in each horde
    a.sort()  # Sort hordes in ascending order to process from largest
    sm = sum(a)  # Total number of monsters
    half = sm // 2  # Target: destroy at least half the monsters using first type of attack
    x = 0  # Counter for second type attacks (ultimate attacks)
    i = n - 1  # Index pointing to the largest remaining horde
    
    # Greedily use first type of attack (normal attack) on hordes
    # until we have enough monsters destroyed to meet the half threshold
    while half > 0:
        half -= a[i]  # Reduce half by monsters in current largest horde
        x += 1  # Increment ultimate attack counter
        i -= 1  # Move to next largest horde
    
    # Result = (monsters destroyed via normal attacks) + (ultimate attacks) + (remaining monsters if odd)
    print(sm // 2 + x + sm % 2)


# https://github.com/VaHiX/CodeForces/